### PR TITLE
examples: OpenShift/OKD template

### PR DESCRIPTION
**Pull request by sapkra**
I'm using Jitsi with OpenShift for some time and contributing my template/manifests now.

My environment:
OpenShift/OKD 3.11

FYI: This PR is currently based on #192.

### CODE REVIEW

**Comment by saghul** [comment]
Changed the base to master. Will land once #192 does.

**Comment by vpavlin** [comment]
@sapkra  This is awesome:) We are looking at Jitsi and to run it on OpenShift as well. Let me know if I can provide any help

**Comment by sapkra** [comment]
@vpavlin Thanks. I definitely need help in testing everything and also find a way to fix jibri recording without being a root user. See my TODOs in the code of #192. Would appreciate your help.

**Comment by Bengrunt** [comment]
Hello @sapkra 

I run an OKD 3.11 cluster with a fairly standard configuration.

I gave a try to your OKD deployment from this pull request and I ran in two issues for the moment:
 - A permission issues with the init script regarding s6 directory creation:

```s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied```

This issue occurs on every image which is logical since I assume it comes from the base image.

- A likely templating issue because when I run the following command :
```
oc process -f jitsi.yaml --ignore-unknown-parameters --param-file ../../.env | oc apply -f -
```

I get most of the resources created except for one deployconfig:
```
Error from server (BadRequest): DeploymentConfig in version "v1" cannot be handled as a DeploymentConfig: v1.DeploymentConfig.Spec: v1.DeploymentConfigSpec.Triggers: v1.DeploymentTriggerPolicies: Template: v1.PodTemplateSpec.Spec: v1.PodSpec.RestartPolicy: Containers: decode slice: expect [ or n, but found {, error found in #10 byte of ...|urces":{}},"restartP|..., bigger context ...|odSeconds":10,"timeoutSeconds":60},"resources":{}},"restartPolicy":"Always"}},"triggers":[{"type":"C|...
```
My `.env` file is quite simple. I mostly set the `DOMAIN` variable as suggested and modified `JVB_PORT` and related variables as well.

Thanks a lot for all this work anyway, that's great and promising!

**Comment by sapkra** [comment]
Thanks for testing! 👍 

For the permission problems you have to build your own images using my changes of #192 util it has been merged. These changes should fix them all.

I will have a look at the template error and will fix it soon. Thanks for pointing this out.

**Comment by Bengrunt** [comment]
> 
> 
> Thanks for testing! 👍
> 
> For the permission problems you have to build your own images using my changes of #192 until it has been merged. These changes should fix them all.
> 
> I will have a look at the template error and will fix it soon. Thanks for pointing this out.

Damn, rookie mistake. Thanks for pointing this out! :)

**Comment by vpavlin** [comment]
FYI Templates are being deprecated see: https://docs.openshift.com/container-platform/4.3/release_notes/ocp-4-3-release-notes.html#ocp-4-3-service-catalog-deprecated 

I would recommend switching to something like Kustomize for the future compatibility

**Comment by Bengrunt** [comment]
> 
> 
> FYI Templates are being deprecated see: https://docs.openshift.com/container-platform/4.3/release_notes/ocp-4-3-release-notes.html#ocp-4-3-service-catalog-deprecated
> 
> I would recommend switching to something like Kustomize for the future compatibility

Hello,

At the moment, Openshift Origin (OKD) 3.11 is the only "stable" release of Openshift that can be self hosted, so until OKD 4 can be successfully deployed on premise I'd be glad if the templates version could be kept as well :)

Thanks a lot to @sapkra for his work on this !

I managed to successfully deploy this stack on my 3.11 OKD cluster.
Still working on welcome page customization, authentication and video bridge scaling. 
I won't need the Jibri and jigasi components so I won't be able to test them.
For etherpad, I guess we could just point to any etherpad deployment, right?

**Comment by sag1** [comment]
Hello,

I used your template to deploy on Openshift. Thanks for sharing.

I have a problem when there are more than 3 participants. I see error in chrome console **Bridge Channel send: no opened channel.** In this [post](https://community.jitsi.org/t/bridge-channel-send-no-opened-channel/18765/8), they found the solution by configuring these values in sip-communicator.properties

org.ice4j.ice.harvest.NAT_HARVESTER_LOCAL_ADDRESS=<Local_Adress>
org.ice4j.ice.harvest.NAT_HARVESTER_PUBLIC_ADDRESS=<Public_Adress>
org.jitsi.videobridge.NAT_HARVESTER_LOCAL_ADDRESS=<Local_Adress>
org.jitsi.videobridge.NAT_HARVESTER_PUBLIC_ADDRESS=<Public_Adress>

I added these parameters and configured **hostIP/PodIP** in **public/private** addresses. Nothing works. Anyone faced this issue on Openshift? Any solution ?

**Comment by sapkra** [comment]
Can you check if the service for the JVB was created and which OpenShift version are you using?

**Comment by sag1** [comment]
@sapkra Yes I checked, JVB service is running and the openshift version is 3.11. For information, it's behind corporate proxy and now i am trying to install TURN server

**Comment by Elyytscha** [comment]
> > FYI Templates are being deprecated see: https://docs.openshift.com/container-platform/4.3/release_notes/ocp-4-3-release-notes.html#ocp-4-3-service-catalog-deprecated
> > I would recommend switching to something like Kustomize for the future compatibility
> 
> Hello,
> 
> At the moment, Openshift Origin (OKD) 3.11 is the only "stable" release of Openshift that can be self hosted, so until OKD 4 can be successfully deployed on premise I'd be glad if the templates version could be kept as well :)
> 
> Thanks a lot to @sapkra for his work on this !
> 
> I managed to successfully deploy this stack on my 3.11 OKD cluster.
> Still working on welcome page customization, authentication and video bridge scaling.
> I won't need the Jibri and jigasi components so I won't be able to test them.
> For etherpad, I guess we could just point to any etherpad deployment, right?

running okd 4.4 on proxmox on the latest FHCOS testing branch over here
`Ready    master   16d   v1.17.1   10.197.0.22   <none>        Fedora CoreOS 31.20200407.20.2   5.5.15-200.fc31.x86_64   cri-o://1.17.1`

idk, it feels way more stable then okd 3.11, even cluster updates are working like a charm,
therefore my okd 3.11 before always exploded like a treadmill ..

actually i'm writing an operator over here which will deploy openshift resources for jitsi meet and stumbled upon this as i searched for openshift support for jitsi meet. 
Hope this will be merged soon so i can use the native docker images instead of building them with additional openshift artbitrary user support by myself ..

**Comment by sapkra** [comment]
Hey @Elyytscha Thanks for testing it on OKD 4.4. I'm in contact with a guy from RedHat who also set up jitsi recently on OCP 4.3 and he is also interested in building a go operator. Maybe you are interested in working together because this can be quite difficult to setup scaling videobridge instances. For more information have a look at [this forum post](https://community.jitsi.org/t/scale-videobridge-inside-kubernetes/20527).

And related to managing template instances: I've never added the templates to the OKD instance. I just processed them in the CI and applied directly them. Updating the templates worked like a charm with this method.

**Comment by Elyytscha** [comment]
> Hey @Elyytscha Thanks for testing it on OKD 4.4. I'm in contact with a guy from RedHat (@ktenzer) who also set up jitsi recently on OCP 4.3 and he is also interested in building a go operator. Maybe you are interested in working together because it can be quite difficult to setup scaling videobridge instances. For more information have a look at [this forum post](https://community.jitsi.org/t/scale-videobridge-inside-kubernetes/20527).
> 
> And related to managing template instances: I've never added the templates to the OKD instance. I just processed them in the CI and directly applied them. Updating the templates worked like a charm with this method. This is the same method I've described in the README of the example.

the jvb in my setup is actually non scaleable, maybe i should rethink this, but it should be possible with a statefulset 

what i mean is if you have 70 jitsi instances in openshift up and running and you want to update the template, you have to reapply all 70 of them manual (or ci triggerd, doesnt matter), so you have to run 70x `oc process -f jitsi.yaml --ignore-unknown-parameters --param-file ../../.env | oc apply -f -`

while an operator is able to manage this with only reapplying the operator which then itself rolls out the patches for the instances.
But it totally agree that doing 70x oc command's is less complex, then writing golang code which manages your application state

**Comment by sapkra** [comment]
I'm running a custom deployment which just have one production, one staging and multiple dev instances. They should only be changed when my template has changed in the specific git branch right now. So if a change requires some other changes or breaks something it is not broken on production. It's a bit hard to explain because our setup is a bit more complex.

To put the JVB in a StafulSet is an interesting idea. But is it possible to shutdown the container if you want to scale down the JVBs? The containers need to be shouted down so that existing conferences are moved to other bridges and will not get new conferences if I understand it correctly.
I definitely have the need for an Operator too when I need to scale the JVBs dynamically. But I don't have any experience with golang yet.

File: examples/openshift/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+# The REST API has to be activated because it's used by the liveness and readiness probes
+JVB_ENABLE_APIS=rest
+
+# The JVB ports have to be 30000 or higher because NodePort is used for them
+JVB_PORT=30000
+JVB_TCP_PORT=30000
+
+# The secrets for all services
+# They can also be generate using the gen-passwords.sh script
```
Comment (ldomesjo) [inline examples/openshift/README.md:19]:
```suggestion
# They can also be generated using the gen-passwords.sh script
```

File: examples/openshift/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+JICOFO_COMPONENT_SECRET=
+JICOFO_AUTH_PASSWORD=
+JVB_AUTH_PASSWORD=
+JIGASI_XMPP_PASSWORD=
+JIBRI_RECORDER_PASSWORD=
+JIBRI_XMPP_PASSWORD=
+```
+
+## Build
+To build the template you have to use use [spiff++](https://github.com/mandelsoft/spiff). It combines all the manifests into one OpenShift Template.
```
Comment (ldomesjo) [inline examples/openshift/README.md:29]:
```suggestion
To build the template you have to use [spiff++](https://github.com/mandelsoft/spiff). It combines all the manifests into one OpenShift Template.
```

File: examples/openshift/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+JIGASI_XMPP_PASSWORD=
+JIBRI_RECORDER_PASSWORD=
+JIBRI_XMPP_PASSWORD=
+```
+
+## Build
+To build the template you have to use [spiff++](https://github.com/mandelsoft/spiff). It combines all the manifests into one OpenShift Template.
+
+```console
+make build
```
Comment (ldomesjo) [inline examples/openshift/README.md:32]: Maybe it's apparent to everyone else but I believe this needs clarification that you need to run make build from the examples/openshift sub-directory as I was first doing from the root directory :)

File: examples/openshift/template.yaml [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+- name: JVB_TCP_PORT
+  required: true
+- name: JVB_BREWERY_MUC
+  required: true
+- name: JVB_AUTH_USER
+  required: true
+- name: JVB_AUTH_PASSWORD
+  required: true
+- name: JVB_STUN_SERVERS
+  required: true
```
Comment (ldomesjo) [inline examples/openshift/template.yaml:80]: I would suggest settings this to required: false as I was having issues with it when running this template in a openshift cluster which does not have access to the internet, the pods were unable to reach the stun server and as such refused to start up, had to disable this in openshift as it is currently not possible to do it via the .env-file

Comment (sapkra) [reply examples/openshift/template.yaml]: I removed this line. If you think this is everything which has to be done, feel free to resolve this review conversation.

File: examples/openshift/README.md (hunk 3 above)
Comment (sapkra) [reply examples/openshift/README.md:36]: I applied your suggested change. :)

**Comment by sapkra** [comment]
Closing it because it would make more sense to use Helm 3 instead.

**Pull request closed without merging**